MSVS CI jobs reorganization

vadz opened the pull request:
Remove MSVS 2019 CI jobs from GitHub Actions as windows-2019 image is being retired.

vadz committed:
Stop on warnings in MSVS CI jobs

Somewhat confusingly, using /WX compiler switch doesn't work, but using
MSBuild /WarnAsError one does.

vadz committed:
Leave only MSVS 2022 CI jobs

MSVS 2019 is only available in windows-2019 image which is being retired
by GitHub, see https://github.com/actions/runner-images/issues/12045

vadz committed:
Remove MSVS 2022 builds from AppVeyor CI

GitHub Actions CI still has those, but doesn't have MSVS 2019 ones any
more, so build wxMSW using this compiler under AppVeyor.

Leave just one build instead of 2 we had before in an attempt to reduce
the total AppVeyor build time.

vadz committed:
Move wxQt/MSVS CI job from AppVeyor to GitHub Actions

It should run faster there, reducing total CI run time.

vadz committed:
Add MSVS wxMSW build using C++20

It seems useful to test the latest support C++ standard version too.

Add cxx20.props which can also be used locally in a similar way, i.e.
after copying it to wx_local.props.

vadz committed:
Use better names for the other MSW CI jobs

Make them more consistent with wxMSW MSVS job.

vadz committed:
Avoid warning about possibly uninitialized variable in a test

MSVS warns about "width" being possibly uninitialized, so do initialize
it and also initialize "height" for consistency.

vadz committed:
Avoid warning about combining enum values form different enums

Cast border style to int explicitly.

vadz committed:
Use more consistent CMake wxMSW CI job name

wxQt job uses wxQt in its name, so use wxMSW for this one.

vadz committed:
Don't install Qt in CMake wxMSW CI job

It's only needed for wxQt, so don't waste time doing it in wxMSW job.

vadz commented:
Hmm, so I've added a wxQt5 build to `ci_msw.yml` without noticing that there was already a wxQt6 build in `ci_cmake.yml`. I should probably move this build to `ci_cmake.yml` too, but what about "msw-msys2-clang" build which also uses CMake, should it be moved as well? Logically it makes sense, but could this make CMake workflow take longer? Or it doesn't matter how we split our jobs between workflows? I'm not really sure how does this work on GitHub, so any advice would be welcome.

MaartenBent commented:
I think all jobs run in parallel, so it doesn't matter how many jobs are added to the matrix in one CI file. 
https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#defining-the-maximum-number-of-concurrent-jobs

It would indeed make sense to move `msw-msys2-clang` to `ci_cmake`, and use `ci_msw` only for the Visual Studio solutions.
Are `makefile.gcc`/`makefile.vc` tested anywhere? Maybe those can be in `ci_msw` too.

vadz commented:
I think we need to get rid of `makefile.{vc,gcc}` anyhow and especially of the VC ones as this is the worst way to build wx: there is no parallelism in nmake, so it takes much longer. And while GNU make is more or less fine, we have a long standing problem with `-j`-unsafed setup.h target... So I am not sure we need CI builds for them or, at least, not ones running for every push/PR — but it could be nice to run them manually to check that they still work before the release.

I'll move the CMake builds as suggested, but I probably won't have time to do it before the end of the week. If you'd like to do it sooner, please feel free to make your own PR replacing this one, just please let me know if you decide to do it. TIA!

vadz committed:
Move wxMSW/clang CI job to the workflow using CMake

This job uses CMake and not MSBuild, so move it to the appropriate
workflow.

vadz committed:
Move wxQt5 CI job to CMake workflow too

For the same reasons as in the parent commit message, this job belongs
to the workflow in ci_cmake.yml rather than MSBuild one in ci_msw.yml.

vadz committed:
Only use "-jN" with make-based CMake generators

Don't use this option with e.g. MSBuild which doesn't understand it.

MaartenBent commented:
I checked past actions and found at least these two that are using `C:/Program Files/LLVM/bin/clang.exe`:
https://github.com/wxWidgets/wxWidgets/actions/runs/15714441700/job/44280535223
https://github.com/wxWidgets/wxWidgets/actions/runs/15713446147/job/44277197704

All the CRT warnings are probably because it uses msvc win32 sdk instead of the msys2/mingw sdk.

One error is `-Wunused-local-typedef`, even though it has `WX_ATTRIBUTE_UNUSED`. It seems the native clang doesn't define `__GNUC__` so `WX_ATTRIBUTE_UNUSED` is empty instead of `__attribute__ ((unused))`.

The other error is `-Wunused-variable`, which is enclosed in a `#if defined(__VISUALC__)` segment. Which is not defined in mingw.

I guess native clang, and probably clang-cl too, define themselves as `__clang__` and `__VISUALC__` instead of `__clang__` and `__GNUC__`.

vadz commented:
Thanks for looking into this, Maarten! I still didn't have time to actually check/confirm this, but your hypothesis makes sense and if native clang doesn't define `__GNUC__`, we really need to review our checks for clang to not rely on it being detected as gcc. This is obviously not urgent, but Id' actually like to be able to use it under Windows, so I've opened #25581 for this.
